Matrices : Faster matrices creation

sylee957 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
#16374

#### Brief description of what is fixed or changed

There are some slowdowns in creating the matrix, and I think one is from `sympify`
```python
from sympy import *
from pyinstrument import Profiler

profiler = Profiler()
profiler.start()

MutableDenseMatrix(1000, 1000, lambda i, j: i+j)

profiler.stop()

print(profiler.output_text(unicode=True, color=True))
```
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 04:48:46  Samples:  4116
 /_//_/// /_\ / //_// / //_'/ //     Duration: 6.645     CPU time: 6.609
/   _/                      v3.0.1

6.655 run_code  IPython\core\interactiveshell.py:3259
└─ 6.655 <module>  <ipython-input-2-3baea188ae1d>:7
   └─ 6.608 __new__  sympy\matrices\dense.py:432
      └─ 6.608 _new  sympy\matrices\dense.py:435
         └─ 6.591 _handle_creation_inputs  sympy\matrices\matrices.py:2287
            └─ 6.531 <listcomp>  sympy\matrices\matrices.py:2483
               ├─ 3.578 sympify  sympy\core\sympify.py:78
               │  ├─ 2.023 [self]  
               │  ├─ 1.436 wrapper  sympy\core\cache.py:92
               │  │  ├─ 1.025 __new__  sympy\core\numbers.py:1930
               │  │  │  ├─ 0.750 __new__  sympy\core\basic.py:96
               │  │  │  └─ 0.275 [self]  
               │  │  └─ 0.411 [self]  
               │  └─ 0.120 __hash__  sympy\core\numbers.py:2128
               ├─ 1.781 <lambda>  <ipython-input-2-3baea188ae1d>:7
               │  ├─ 1.603 __add__  sympy\core\numbers.py:2003
               │  │  ├─ 1.036 wrapper  sympy\core\cache.py:92
               │  │  │  ├─ 0.743 __new__  sympy\core\numbers.py:1930
               │  │  │  │  ├─ 0.457 [self]  
               │  │  │  │  └─ 0.286 __new__  sympy\core\basic.py:96
               │  │  │  └─ 0.293 [self]  
               │  │  └─ 0.567 [self]  
               │  └─ 0.178 [self]  
               └─ 1.171 [self]  
```
I have replaced `sympify` with `_sympify`, regarding the comment in https://github.com/sympy/sympy/issues/15416#issuecomment-453690358
~~and also removed sympifying the index, which is not needed for lambdas written in pure python integer operations.~~
and also made it to iterate over already simpified list, which will make `sympify` call `M + N` times rather than calling sympify `M * N` times while iterating over all the matrix elements as previous.

Things seems like going faster with the creation input of index and lambda.
But I think that it is more contributed by the latter change, rather than the prior change though

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 14:10:44  Samples:  1509
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.177     CPU time: 2.156
/   _/                      v3.0.1
```

I had confirmed that Sparse Matrix follows the same speedup.
and it may need some performance investigation for, other ways of matrix creation from list or such.

#### Other comments

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- matrices
  - Improved speed for matrix constructor using rows, cols, and function.
<!-- END RELEASE NOTES -->

sylee957 committed:
Remove sympify for indexing

smichr commented:
```python
Failed example:
    Matrix(2, 2, lambda i, j: i/(j + 1) )
Expected:
    Matrix([
    [0,   0],
    [1, 1/2]])
Got:
    Matrix([
    [0.0, 0.0],
    [1.0, 0.5]])
```
So it looks like the lambda is being fed ints. But below it looks like the range values are being sympified so perhaps somewhere else this is not happening.

https://github.com/sympy/sympy/blob/b176f6a1d9890b42dc361857c887992315e3d5ad/sympy/matrices/matrices.py#L2478-L2484

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2480,7 +2481,7 @@ def do(x):
                 for i in range(rows):
                     flat_list.extend(
-                        [cls._sympify(op(cls._sympify(i), cls._sympify(j)))
+                        [cls._sympify(op(i, j))
```
```suggest
                        [op(cls._sympify(i), cls._sympify(j))
```

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2480,7 +2481,7 @@ def do(x):
                 flat_list = []
                 for i in range(rows):
```
```python
                for i in range(rows):
                    i = _sympify(i)
                    flat_list.extend([op(i, cls._sympify(j)) for j in range(cols)])
```
or
```python
                J = [_sympify(j) for j in range(cols)]
                for i in range(rows):
                    i = _sympify(i)
                    flat_list.extend([op(i, J[j]) for j in range(cols)])
```

sylee957 committed:
Sympify the iterables only once

sylee957 committed:
Fix attributeerror

sylee957 commented:
I see the failing tests are because previous implementation allowed creating matrices from a 3D array, strings or dictionaries or such.

Should we disallow these type of matrices?
```python
from sympy import *
F, G = Function('F'), Function('G')

Matrix([["123"]])
Matrix([[[1, 2, 3], [4, 5, 6]]])
Matrix(2, 2, 
    [
        (G(0), {F(0): G(0)}),
        (G(1), {F(1): G(1)}),
        (G(1), {F(1): G(1)}),
        (G(2), {F(2): G(2)})
    ]
)
```

oscarbenjamin commented:
I don't understand what the tuple-dict matrix is:
```
In [11]: M = Matrix(2, 2,  
    ...:     [ 
    ...:         (G(0), {F(0): G(0)}), 
    ...:         (G(1), {F(1): G(1)}), 
    ...:         (G(1), {F(1): G(1)}), 
    ...:         (G(2), {F(2): G(2)}) 
    ...:     ] 
    ...: )                                                                                                                                                                                                         

In [12]: M                                                                                                                                                                                                         
Out[12]: 
⎡(G(0), {F(0): G(0)})  (G(1), {F(1): G(1)})⎤
⎢                                          ⎥
⎣(G(1), {F(1): G(1)})  (G(2), {F(2): G(2)})⎦

In [14]: M+M                                                                                                                                                                                                       
Out[14]: 
⎡(G(0), {F(0): G(0)}, G(0), {F(0): G(0)})  (G(1), {F(1): G(1)}, G(1), {F(1): G(1)})⎤
⎢                                                                                  ⎥
⎣(G(1), {F(1): G(1)}, G(1), {F(1): G(1)})  (G(2), {F(2): G(2)}, G(2), {F(2): G(2)})⎦

In [16]: 3*M                                                                                                                                                                                                       
Out[16]: 
⎡(G(0), {F(0): G(0)}, G(0), {F(0): G(0)}, G(0), {F(0): G(0)})  (G(1), {F(1): G(1)}, G(1), {F(1): G(1)}, G(1), {F(1): G(1)})⎤
⎢                                                                                                                          ⎥
⎣(G(1), {F(1): G(1)}, G(1), {F(1): G(1)}, G(1), {F(1): G(1)})  (G(2), {F(2): G(2)}, G(2), {F(2): G(2)}, G(2), {F(2): G(2)})⎦
```
Is there a use for that somewhere in SymPy?

smichr commented:
> Is there a use for that somewhere in SymPy?

https://github.com/sympy/sympy/pull/16462#issuecomment-481052099

Maybe we need to use TableForm or something like a 2d Tuple for such things rather than allowing Matrix to be used for such Purposes. Maybe a TableMatrix class?

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2479,10 +2479,11 @@ def do(x):
                 flat_list = []
-                for i in range(rows):
+                rows_range = [cls._sympify(i) for i in range(rows)]
+                cols_range = [cls._sympify(j) for j in range(cols)]
```
or `ix = [cls._sympify(i) for i in range(max(row, cols))]` and then index `ix[i]` or `ix[j]` as needed.

oscarbenjamin commented:
If you have tuple elements then I presume you are not intending to use Matrix operations in which case why not just a list of lists (or tuple of tuples)?

smichr commented:
> why not just a list of lists

good question...ability to present it nicely is the only advantage I see. I didn't look much into how (or why) it was used in the reference I gave.

sylee957 commented:
Things like `(G(0), {F(0): G(0)})` should better belong to some tabular database, rather than matrix.
I guess it's used for the key `map` in `replace`, but it can be a wrong result, and I think it should rather give something like `(Matrix([[G(0), G(1)], [G(1), G(2)]]), {F: G})` rather than some tuples and dictionaries, to be valid.

Since we already flatten block matrix like after #16462,
```python
from sympy import *

A = Matrix([[1, 2], [3, 4]])
B = Matrix([[5, 6], [7, 8]])
C = Matrix([[9, 10], [11, 12]])
D = Matrix([[13, 14], [15, 16]])

Matrix([[A, B], [C, D]])
```
 I don't think there should be any reason to allow arrays inside the matrix. If it should, it should use `NDimArray` for uniform array types, or plane python lists if it is ragged array.

oscarbenjamin commented:
@sylee957 I'm not quite sure what you are suggesting.

I agree that matrix elements should at least be Expr. In linear algebra matrix elements are generally from at least a ring if not a field (the main non-field example is matrices as elements in block matrices).

sylee957 commented:
For the block matrix, I mean the consistency between these

```python
>>> from sympy import *

>>> A = Matrix([[1, 2], [3, 4]])
>>> B = Matrix([[5, 6], [7, 8]])
>>> C = Matrix([[9, 10], [11, 12]])
>>> D = Matrix([[13, 14], [15, 16]])

>>> Matrix([[A, B], [C, D]])
Matrix([
[ 1,  2,  5,  6],
[ 3,  4,  7,  8],
[ 9, 10, 13, 14],
[11, 12, 15, 16]])
```

```python
>>> Matrix(
...     [
...         [
...             ((1, 2), (3, 4)),
...             ((5, 6), (7, 8))
...         ],
...         [
...             ((9, 10), (11, 12)),
...             ((13, 14), (15, 16))
...         ]
...     ]
... )
Matrix([
[   ((1, 2), (3, 4)),     ((5, 6), (7, 8))],
[((9, 10), (11, 12)), ((13, 14), (15, 16))]])
```

smichr commented:
How hard is it to rewrite `laplace_transforms` to use `NDimArray` or alternative?

sylee957 commented:
I see `laplace_transform` can accept matrix as argument
https://github.com/sympy/sympy/blob/b8fe457a02cc24b3470ff678d0099c350b7fef43/sympy/integrals/transforms.py#L1135-L1136

<img src=https://user-images.githubusercontent.com/34944973/57059322-c33c4000-6cef-11e9-955d-068a35ae95ea.png width=256px>

Personally I didn't like the integral transformations giving convergence condition as a tuple, rather than inside a `Piecewise`, which had already been inconsistent with `integrate` itself, and in this case, it may have to force to `Piecewise` if used inside matrix. It may be time to make such propositions.

Also, how would Laplace transformation for matrices be defined mathematically.
I have found some example of inverse Laplace transformation used in computing matrix exponential, like above example though, but how would Laplace transform for matrix-valued variables generalized as elementwise laplace transformation? Can it be possible for non-square matrices?

oscarbenjamin commented:
Laplace transforms for matrices are well defined. As for any matrix you can integrate elementwise. The problem above is that the tuples (expr, cond) are inside the matrix like
```

In [25]: Matrix([[(1, True), (2, True)], [(3, True), (4, True)]])                                                                 
Out[25]: 
⎡(1, True)  (2, True)⎤
⎢                    ⎥
⎣(3, True)  (4, True)⎦
```
 It should be a single matrix with expression elements. That can be returned as a condition tuple as in
```
In [23]: (Matrix([[1, 2], [3, 4]]), True)                                                                                         
Out[23]: 
(Matrix([
 [1, 2],
 [3, 4]]), True)
```
I guess the condition should be the logical and of the conditions for the elements.

sylee957 commented:
I think `laplace_transform` is rather giving 3-tuple instead of 2-tuple.
And also, I find that laplace transformation can fail for some matrix elements, yielding unevaluated laplace transform.
And in this case, how would it affect the convergence condition for the matrix as a whole, if something fails inside. 

Personally, it is still complicated to clean up some complications, besides the other issue with `replace` returning dictionaries.
Also, Things related to `sympify` and `replace` and `laplace_transform` can be backward incompatible.

I would rather make `sympify` stay same as it was for now. And would want to merge the changes regarding cacheing the indices, because it contributes to the speed improvement, mostly for now.
And probably make the discussion separate, for integral transformations.

smichr committed:
Only sympify needed ints once

smichr committed:
only sympify needed ints once

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2478,10 +2478,14 @@ def do(x):
                     flat_list.extend(
-                        [cls._sympify(op(cls._sympify(i), cls._sympify(j)))
-                         for j in range(cols)])
+                        [cls._sympify(op(i, j)) for j in cols_range])
```
Do you think a warning not to calculate non-Sympy values could eliminate the need of this additional `_sympify`? Any operation done with the Ints will result in an Expr ... unless the user purposely does something like `int(i/(j + 1))` instead of `Integer(i/(j + 1))` in the lambda.

sylee957 replied:
I have experimented with 
1. Passing SymPy integer to the callback and not sympifying the collected results.
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:59:55  Samples:  816
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.209     CPU time: 1.188
/   _/                      v3.0.1
```

2. Passing python integer to the callback and sympifying the collected results
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:56:25  Samples:  931
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.456     CPU time: 1.453
/   _/                      v3.0.1
```

And both gave more faster matrix creation than
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 16:08:17  Samples:  1423
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.078     CPU time: 2.062
/   _/                      v3.0.1
``` 

I think that the arguments passed to, or passed by the callback would not necessarily have to be sympified, if the operation has no precision issues.
But before writing the `flat_list`, the input types have to be sympified can be a must.

And the first example hadn't been tested with the type checking, so it would possibly add more time for the type checking. And I'm afraid that threading through all the sympy types would not be much different than using `sympify`

So, personally, I think it would be better to make a new argument to specify whether the indices should be python integer or sympy integer.

sylee957 replied:
I have experimented with 
1. Passing SymPy integer to the callback and not sympifying the collected results.
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:59:55  Samples:  816
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.209     CPU time: 1.188
/   _/                      v3.0.1
```

2. Passing python integer to the callback and sympifying the collected results
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:56:25  Samples:  931
 /_//_/// /_\ / //_// / //_'/ //     Duration: 1.456     CPU time: 1.453
/   _/                      v3.0.1
```

And both gave more faster matrix creation than
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 16:08:17  Samples:  1423
 /_//_/// /_\ / //_// / //_'/ //     Duration: 2.078     CPU time: 2.062
/   _/                      v3.0.1
``` 

I think that the arguments passed to, or passed by the callback would not necessarily have to be sympified, if the operation has no precision issues.
But before writing the `flat_list`, the input types have to be sympified can be a must.

And the first example hadn't been tested with the type checking, so it would possibly add more time for the type checking. And I'm afraid that threading through all the sympy types would not be much different than using `sympify`

So, personally, I think it would be better to make a new argument to specify whether the indices should be python integer or sympy integer.

sylee957 commented on sympy/matrices/matrices.py:
```diff
@@ -2478,10 +2478,14 @@ def do(x):
+                if rows < cols:
+                    rows_range = rows_range[:rows]
+                elif cols < rows:
+                    cols_range = cols_range[:cols]
```
I think that the performance improvement is not visible for the last changes, from my experiment,
Probably because of the nature that linear time complexity does not contribute significantly to the matrix.
But I'm keeping this because there was this specific part is faster for creating a list with a million elements.

smichr replied:
How about a *single flag* that controls only *output*:
`sympify_output` (which defaults to False).
```
ix = [_sympify(i) for i in range(max(rows, cols))]
rows_range = ix[:rows_range]
cols_range = ix[:cols_range]
for i in rows_range:
    flat.extend([f(i,j) for j in cols_range])
if sympify_output:
    flat = map(_sympify,flat)
```

sylee957 committed:
Add new argument

sylee957 committed:
Add flag for sympify

sylee957 committed:
Add flag for sympify in sparse matrix

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2508,9 +2520,15 @@ def do(x):
-                    flat_list.extend(
-                        [cls._sympify(op(i, j)) for j in cols_range])
+
+                if sympify_callback_return:
```
recommend
``python
for i in rows_range:
    flat_list.extend([op(i, j) for j in cols_range])
if sympify_callback_return:
    flat_list = cls._sympify(flat_list)
```
And I highly recommend just dropping the option to sympify the input because, as you point out, this is only an `O(n)` step in an `O(n**2)` process.

sylee957 replied:
Well, I think that things have to be done like either one of these
1. Sympify first and if things are keeping sympy context, it doesn't necessarily have to go through `o(n**2)` sympifying process
2. Sympify last and it doesn't necessarily have to go through `o(n)` index sympifying process,
3. If we make two flag `False`, sympy types should be created inside the callback function

And the third one has significant slowdown like the previous version than this, because from the mapped function, integer objects should be created `n^2` times, and this was why `Matrix` constructor can take advantage of storing pre-made things in a vector.

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2301,7 +2301,17 @@ def _handle_creation_inputs(cls, *args, **kwargs):
+
+            If ``False``, it skips the process.
+
+            This flag is unsafe for use if the input is not guaranteed
```
It's the output that has to be guaranteed to be SymPy sanitized, not the input:

perhaps this, instead:

Setting this to False can safely be done if one is not returning int values from 
the callback functon as in `lambda i, j: int(i + j)` which would produce a Python int
instead of `lambda i, j: i + j` which would produce a SymPy Integer.

sylee957 committed:
Revise the docs

sylee957 committed:
Add tests

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -2288,6 +2288,35 @@ def irregular(cls, ntop, *matrices, **kwargs):
+
+            If ``False``, it skips the process.
+
+            This flag is unsafe for use if the callabck returns python
```
callabck -> callback
python -> Python

The reason I didn't mention float is that any float that is encountered will just get recast to Float when needed. The only danger (and potentially hard to track down bug) is if ints are used with non-future semantics and truncate during division.

```
This flag is unsafe for use if the callback returns non-native
SymPy types. For example, if Python ``ints``  are returned and not sympified
they may
truncate during computation of the inverse of a matrix. In addition, any
non-native SymPy object will not have the methods that
SymPy expects them to have.
```

oscarbenjamin commented:
What's the status of this?

sylee957 committed:
Add volatile constructor from flattened list

sylee957 committed:
Rearrange some creation methods

Split creation methods

Split creation routine

Add developers reference doc and fix py2 issues

Fix doctests

Add documentation about creation handler from matrixlike elements

Add some examples

Add parameters section

Fix pop from empty set error

Add an example about empty matrix

Revise doctest and refactor attribute

Fix doctest

sylee957 committed:
Resolve merge conflict

sylee957 committed:
Remove redundant check

smichr commented on sympy/matrices/tests/test_matrices.py:
```diff
@@ -330,6 +330,65 @@ def test_creation():
+            ).tolist() == \
+                [[S(0), S(1)], [S(1), S(2)]]
+
+    assert Matrix(
```
```python
from sympy.core.basic import _aresame
assert _aresame(Matrix(2, 2, lambda i, j: i + j,
            sympify_callback_arguments=False,
            sympify_callback_return=False
            ).tolist(),
                [[0, 1], [1, 2]])
```

The `==` will not affirm that these values are not Integers. You could use it on all these assertions.

sylee957 commented:
While doing some cleanups for handle_creation_input,
I've found some other inconsistency like

```python
>>> from sympy import *
>>> Matrix([[1, ones(1, 1)]])
Matrix([[1, Matrix([[1]])]])
>>> Matrix([1, ones(1, 1)])
Matrix([
[1],
[1]])
```

sylee957 committed:
Rename methods

sylee957 committed:
Add parameters section

sylee957 committed:
Add examples

sylee957 committed:
Add row-vector handler

sylee957 committed:
Remvoe evaluate tag

sylee957 committed:
Fix flattening

sylee957 committed:
Use _mat

sylee957 committed:
Add row vector handler

sylee957 committed:
Rename to column vector

sylee957 committed:
Remove redundant creation

sylee957 committed:
Remove redundant handler

sylee957 committed:
Add evaluate

sylee957 committed:
Add sympify

sylee957 committed:
Allow empty container

sylee957 committed:
Fix tests
